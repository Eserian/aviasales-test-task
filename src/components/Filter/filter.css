.filter {
    background: #fff;
    border: 1px solid #DFE5EC;
    box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.1);
    border-radius: 5px;
    text-align: left;
    height: 250px;
}

.col-4 {
    grid-column: span 4;
}

.filter-header {
    font-size: 12px;
    line-height: 12px;
    letter-spacing: 0.5px;
    text-transform: uppercase;
    color: #4A4A4A;
    padding: 20px 20px 5px;
}

.checkbox-list {
    font-size: 13px;
    font-weight: normal;
    line-height: 20px;
}



