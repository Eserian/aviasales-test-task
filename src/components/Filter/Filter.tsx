import React, { FC, useState, useCallback }from 'react';
import './filter.css';
import { Checkbox } from '../Checkbox/Checkbox';

type stops = {
  [key: string] : { label: string,  checked: boolean, filterParam: number }
};

type filterProps = {
  handleFilter: (filterParams: number[]) => void
}

export const Filter: FC<filterProps> = ({handleFilter}) => {

  const initStops: stops = {
    'all': { label: 'Все', checked: true, filterParam: -1 },
    'without stops': { label: 'Без пересадок', checked: true, filterParam: 0 },
    'one stop': { label: '1 пересадка', checked: true, filterParam: 1 },
    'two stops': { label: '2 пересадки', checked: true, filterParam: 2 },
    'three stops': { label: '3 пересадки', checked: true, filterParam: 3 }
  };

  const [stops, setStops] = useState(initStops);

  const handleStopsChange = useCallback((e: any) => {
    const stopType: string = e.target.dataset.type;
    const isChecked: boolean = e.target.checked;
    
    if (stopType === 'all') {
      setStops(Object.entries(stops).reduce((acc, [key, value]) => ({ ...acc, [key]: { ...value, checked: isChecked } }), {}));
      handleFilter(isChecked ? [0, 1, 2, 3] : []);
      return;
    }

    const newStops = { ...stops, [stopType]: { ...stops[stopType], checked: isChecked }}
    const isAllFieldsChecked: boolean = Object.entries(newStops).filter(([key, ]) => key !== 'all').every(([, { checked }]) => checked);
    newStops['all'] = { ...newStops['all'], checked: isAllFieldsChecked };
    setStops(newStops);
    handleFilter(Object.keys(newStops).filter((key) => newStops[key].checked).map((key) => newStops[key].filterParam));
  }, [handleFilter, stops]);

  return (
    <aside className="filter col-4">
      <div className="filter-header">Количество пересадок</div>
      <div className="checkbox-list">
        {
          Object.entries(stops).map(([key, { label, checked }], i) => (
            <Checkbox
              key={i}
              label={label}
              stopType={key}
              isChecked={checked}
              onChange={handleStopsChange}
            />
          ))
        }
      </div>
    </aside>
  )
};
