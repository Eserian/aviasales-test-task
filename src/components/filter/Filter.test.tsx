import React from 'react';
import { render } from '@testing-library/react';
import { Filter } from './Filter';

test('render correctly', () => {
  const handleStopsChange = () => {};
  const stops = {
    0: true,
    1: true,
    2: true,
    3: true,
  };

  const filter = render(<Filter stops={stops} handleStopsChange={handleStopsChange} />);
  expect(filter).toMatchSnapshot();
});
