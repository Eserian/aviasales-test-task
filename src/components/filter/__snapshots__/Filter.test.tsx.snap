// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <aside
        class="filter col-4"
      >
        <div
          class="filter-header"
        >
          Количество пересадок
        </div>
        <div
          class="checkbox-list"
        >
          <div
            class="checkbox"
          >
            <label
              class="checkbox-label"
            >
              <input
                checked=""
                class="checkbox-input"
                data-type="all"
                type="checkbox"
              />
              <span
                class="checkbox-display"
              />
              Все
            </label>
          </div>
          <div
            class="checkbox"
          >
            <label
              class="checkbox-label"
            >
              <input
                checked=""
                class="checkbox-input"
                data-type="0"
                type="checkbox"
              />
              <span
                class="checkbox-display"
              />
              Без пересадок
            </label>
          </div>
          <div
            class="checkbox"
          >
            <label
              class="checkbox-label"
            >
              <input
                checked=""
                class="checkbox-input"
                data-type="1"
                type="checkbox"
              />
              <span
                class="checkbox-display"
              />
              1 пересадка
            </label>
          </div>
          <div
            class="checkbox"
          >
            <label
              class="checkbox-label"
            >
              <input
                checked=""
                class="checkbox-input"
                data-type="2"
                type="checkbox"
              />
              <span
                class="checkbox-display"
              />
              2 пересадки
            </label>
          </div>
          <div
            class="checkbox"
          >
            <label
              class="checkbox-label"
            >
              <input
                checked=""
                class="checkbox-input"
                data-type="3"
                type="checkbox"
              />
              <span
                class="checkbox-display"
              />
              3 пересадки
            </label>
          </div>
        </div>
      </aside>
    </div>
  </body>,
  "container": <div>
    <aside
      class="filter col-4"
    >
      <div
        class="filter-header"
      >
        Количество пересадок
      </div>
      <div
        class="checkbox-list"
      >
        <div
          class="checkbox"
        >
          <label
            class="checkbox-label"
          >
            <input
              checked=""
              class="checkbox-input"
              data-type="all"
              type="checkbox"
            />
            <span
              class="checkbox-display"
            />
            Все
          </label>
        </div>
        <div
          class="checkbox"
        >
          <label
            class="checkbox-label"
          >
            <input
              checked=""
              class="checkbox-input"
              data-type="0"
              type="checkbox"
            />
            <span
              class="checkbox-display"
            />
            Без пересадок
          </label>
        </div>
        <div
          class="checkbox"
        >
          <label
            class="checkbox-label"
          >
            <input
              checked=""
              class="checkbox-input"
              data-type="1"
              type="checkbox"
            />
            <span
              class="checkbox-display"
            />
            1 пересадка
          </label>
        </div>
        <div
          class="checkbox"
        >
          <label
            class="checkbox-label"
          >
            <input
              checked=""
              class="checkbox-input"
              data-type="2"
              type="checkbox"
            />
            <span
              class="checkbox-display"
            />
            2 пересадки
          </label>
        </div>
        <div
          class="checkbox"
        >
          <label
            class="checkbox-label"
          >
            <input
              checked=""
              class="checkbox-input"
              data-type="3"
              type="checkbox"
            />
            <span
              class="checkbox-display"
            />
            3 пересадки
          </label>
        </div>
      </div>
    </aside>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
